druid/src/widget/svg.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An SVG widget.
+
+use crate::{
+    kurbo::BezPath,
+    piet::{Color, RenderContext},
+    BaseState, BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, PaintCtx, Point, Size,
+    UpdateCtx, Widget,
+};
+
+use log::error;
+use std::error::Error;
+use std::marker::PhantomData;
+use std::str::FromStr;
+
+use usvg;
+
+/// A widget that renders a SVG
+pub struct Svg<T> {
+    // On construction the SVG string is parsed into a SvgData.
+    svg_data: SvgData,
+    phantom: PhantomData<T>,
+}
+
+impl<T: Data> Svg<T> {
+    /// Create an SVG-drawing widget from a valid SVG string literal.
+    ///
+    /// The SVG will scale to fit its box constraints.
+    /// If SVG is invalid a blank SVG will be rendered instead.
+    pub fn new(svg_str: &str) -> impl Widget<T> {
+        let svg_data = SvgData::from_str(svg_str);
+
+        match svg_data {
+            Ok(data) => Svg {
+                svg_data: data,
+                phantom: Default::default(),
+            },
+            Err(err) => {
+                error!("{}", err);
```

Comment:
I think this is reasonable until we figure out a better error handling story. 🤷‍♀

Reply:
Is there a good reason not to return `Result`?

Reply:
not a great reason, but my thinking right now is that no other widget constructors are failable, and it feels weird to start here.

My long-term vision for this is that you shouldn't be loading data directly from the file system; there should be a 'resource bundle' that druid exposes that contains various types of resources (image files, font files, localization files) and that is validated at compile time, so that at runtime getting a resource looks something like,

```rust
let my_svg: SvgData = env.get_bundle_resource(MY_SVG_IDENTIFIER);
```

or maybe even in a very fancy world there's a build phase that generates per-app marker 'keys' (like `env::Key`) for each of the resources in this app's bundle?

Reply:
I can open an issue for this if you'd like?

Reply:
sure!